mds: dont print subtrees if they are too big or too many

rishabh-d-dave opened the pull request:
Fixes: https://tracker.ceph.com/issues/37726

rishabh-d-dave commented on src/mds/MDCache.cc:
```diff
@@ -12035,6 +12035,28 @@ void MDCache::show_subtrees(int dbl)
+    dout(15) << "SUBTREES_DEPTH_THRESHOLD = " << SUBTREES_DEPTH_THRESHOLD
+	     << dendl;
+  }
+  //if (current_logging_level < 25)
```
@batrick How do I get the current logging level? I tried `g_conf().get_val<Option::size_t>("debug_mds")` but it didn't work for me.

batrick replied:
See: https://github.com/ceph/ceph/blob/2c6a785410311041a86abdad235ebc3d1830813d/src/mds/Migrator.cc#L622

batrick replied:
`should_gather` returns a boolean indicating whether the log message would be generated. i.e. 

`g_conf()->subsys.should_gather<ceph_subsys_mds, 5>()`

returns true if

`dout(5) << foo << dendl`

would print something.

Or am I misunderstanding you?

rishabh-d-dave replied:
You are right. Initially, I felt that default value of `debug mds` was being returned by `g_conf()->subsys.should_gather<ceph_subsys_mds, 5>()` (since it always returned, as I report in my previous comment, and the default value for `debug mds` is also 1). So, I was looking for another function that would return the current value.

rishabh-d-dave commented on src/mds/MDBalancer.cc:
```diff
@@ -150,6 +150,15 @@ void MDBalancer::handle_export_pins(void)
   set<CDir *> authsubs;
   mds->mdcache->get_auth_subtrees(authsubs);
+
+  if (authsubs.size() > AUTH_TREES_THRESHOLD) {
```
While testing the other commit on this PR, I observed that an "auth tree" was being reported every few seconds; so wrote a similar patch for it too.

For the sake of book-keeping should I create an issue on the tracker?

batrick replied:
It's fine to include it in this PR/issue.

I think even 50 is too high. Maybe 5?

rishabh-d-dave replied:
Would it be nice to check the auth tree's depth too? We are doing it for subtrees but not for auth trees.

batrick replied:
IIUC, the subtree depth doesn't factor in here. `authsubs` contains all authoritative subtrees at all depths.

batrick commented on src/mds/MDBalancer.cc:
```diff
@@ -150,6 +150,15 @@ void MDBalancer::handle_export_pins(void)
+  if (authsubs.size() > AUTH_TREES_THRESHOLD) {
+    dout(15) << "number of auth tree is beyond threshold" << dendl;
+    dout(15) << "number of auth trees = " <<  authsubs.size() << dendl;
+    dout(15) << "AUTH_TREES_THRESHOLD = " << AUTH_TREES_THRESHOLD << dendl;
```
Collapse this into a single short message. These high debug (15) messages do not need to be easily readable. As long as you can find it in the source, it's easy enough to figure it out.

batrick commented on src/mds/MDBalancer.h:
```diff
@@ -82,6 +82,7 @@ class MDBalancer {
 private:
   bool bal_fragment_dirs;
   int64_t bal_fragment_interval;
+  unsigned int AUTH_TREES_THRESHOLD = 50;
```
`static const ...`

batrick commented on src/mds/MDCache.h:
```diff
@@ -303,6 +303,9 @@ class MDCache {
   // -- subtrees --
+private:
+  unsigned int SUBTREES_COUNT_THRESHOLD = 50;
+  unsigned int SUBTREES_DEPTH_THRESHOLD = 50;
```
ditto

rishabh-d-dave commented:
Changes -
* reduce the threshold level to 5
* shortened the log messages
* added `static` and `const` qualifiers to threshold variables
* Added the condition to return if logging level is below 25

batrick commented on src/mds/MDBalancer.h:
```diff
@@ -82,6 +82,7 @@ class MDBalancer {
 private:
   bool bal_fragment_dirs;
   int64_t bal_fragment_interval;
+  const static unsigned int AUTH_TREES_THRESHOLD = 5;
```
`static const...`

batrick commented on src/mds/MDCache.cc:
```diff
@@ -12035,6 +12035,20 @@ void MDCache::show_subtrees(int dbl)
     }
   }
 
+  if (subtrees_seen.size() > SUBTREES_COUNT_THRESHOLD && depth >
```
I think we can be a little smarter and move these checks up to the while loop above so we don't traverse all the subtrees only to do an early return here.

batrick commented on src/mds/MDCache.cc:
```diff
@@ -12035,6 +12035,20 @@ void MDCache::show_subtrees(int dbl)
+  if (subtrees_seen.size() > SUBTREES_COUNT_THRESHOLD && depth >
+      SUBTREES_DEPTH_THRESHOLD) {
+    dout(15) << "number of subtrees > SUBTREES_COUNT_THRESHOLD and depth > "
+		"SUBTREES_DEPTH_THRESHOLD; not printing subtrees" << dendl;
```
Let's at least print something a little useful like  the number of subtrees in this line. I think we can forgo printing "subtrees > SUB..." stuff.

batrick commented on src/mds/MDCache.cc:
```diff
@@ -12035,6 +12035,20 @@ void MDCache::show_subtrees(int dbl)
+  }
+
+  if (!g_conf()->subsys.should_gather<ceph_subsys_mds, 25>())
+    return;
```
Let's add a bool argument that forces the subtrees to be printed. We want the full subtree list probably every ~5 seconds via `MDBalancer::tick`.

rishabh-d-dave replied:
> Let's add a bool argument that forces the subtrees to be printed. 

I don't get this part.

rishabh-d-dave replied:
> > Let's add a bool argument that forces the subtrees to be printed.
> 
> I don't get this part.

Adding a bool variable that would print subtrees irrespective of their number/size. Right?

batrick replied:
> > > Let's add a bool argument that forces the subtrees to be printed.
> > 
> > 
> > I don't get this part.
> 
> Adding a bool variable that would print subtrees irrespective of their number/size. Right?

Yes, subject to the passed in `dbl`. i.e. instead do this:

`if (!force && !g_conf()->subsys.should_gather<ceph_subsys_mds, 25>())`

rishabh-d-dave commented:
@batrick 
I can still see subtrees in logs printed via `create_subtree_map` -
```
2019-01-29 16:23:59.018 7f0db9665700 10 mds.0.cache create_subtree_map 102 subtrees, 102 fullauth
2019-01-29 16:23:59.018 7f0db9665700  0 mds.0.cache debugtask dbl = 10
2019-01-29 16:23:59.018 7f0db9665700  0 mds.0.cache debugtask ceph::dout::need_dynamic(dbl) = 10
2019-01-29 16:23:59.018 7f0db9665700 15 mds.0.cache number of subtrees = 102; not printing subtrees
2019-01-29 16:23:59.018 7f0db9665700 15 mds.0.cache  subtree [dir 0x1 / [2,head] auth{1=1} pv=12433 v=12431 cv=1/1 REP dir_auth=0 ap=1+1 state=1610874881|complete f(v0 m2019-01-29 16:23:03.901419 100=0+100)->f(v0 m2019-01-29 16:23:03.901419 100=0+100) n(v5 rc2019-01-29 16:23:41.593010 b20020 1102=1001+101)->n(v6 rc2019-01-29 16:23:41.593010 b20020 1102=1001+101) hs=100+0,ss=0+0 dirty=100 | child=1 subtree=1 replicated=1 dirty=1 authpin=1 0x55f91c0f4500]
2019-01-29 16:23:59.018 7f0db9665700 15 mds.0.cache  subtree bound [dir 0x10000000000 /dir00/ [2,head] auth v=151 cv=0/0 dir_auth=0 state=1611399169|complete|auxsubtree f(v0 m2019-01-29 16:23:09.459607 11=10+1) n(v3 rc2019-01-29 16:23:41.593010 b220 12=11+1) hs=11+0,ss=0+0 dirty=11 | child=1 subtree=1 dirty=1 authpin=0 0x55f91c175400]
2019-01-29 16:23:59.018 7f0db9665700 15 mds.0.cache  subtree bound [dir 0x10000000001 /dir01/ [2,head] auth v=121 cv=0/0 dir_auth=0 state=1611399169|complete|auxsubtree f(v0 m2019-01-29 16:23:03.968042 10=10+0) n(v0 rc2019-01-29 16:23:10.540687 b200 10=10+0) hs=10+0,ss=0+0 dirty=10 | child=1 subtree=1 dirty=1 authpin=0 0x55f91c175900]
2019-01-29 16:23:59.018 7f0db9665700 15 mds.0.cache  subtree bound [dir 0x10000000002 /dir02/ [2,head] auth v=121 cv=0/0 dir_auth=0 state=1611399169|complete|auxsubtree f(v0 m2019-01-29 16:23:03.999003 10=10+0) n(v0 rc2019-01-29 16:23:10.588399 b200 10=10+0) hs=10+0,ss=0+0 dirty=10 | child=1 subtree=1 dirty=1 authpin=0 0x55f91c175e00]
2019-01-29 16:23:59.018 7f0db9665700 15 mds.0.cache  subtree bound [dir 0x10000000003 /dir03/ [2,head] auth v=121 cv=0/0 dir_auth=0 state=1611399169|complete|auxsubtree f(v0 m2019-01-29 16:23:04.034113 10=10+0) n(v0 rc2019-01-29 16:23:10.632857 b200 10=10+0) hs=10+0,ss=0+0 dirty=10 | child=1 subtree=1 dirty=1 authpin=0 0x55f91c176300]
2019-01-29 16:23:59.018 7f0db9665700 15 mds.0.cache  subtree bound [dir 0x10000000004 /dir04/ [2,head] auth v=121 cv=0/0 dir_auth=0 state=1611399169|complete|auxsubtree f(v0 m2019-01-29 16:23:04.066848 10=10+0) n(v0 rc2019-01-29 16:23:10.670789 b200 10=10+0) hs=10+0,ss=0+0 dirty=10 | child=1 subtree=1 dirty=1 authpin=0 0x55f91c176800]
2019-01-29 16:23:59.018 7f0db9665700 15 mds.0.cache  subtree bound [dir 0x10000000005 /dir05/ [2,head] auth v=121 cv=0/0 dir_auth=0 state=1611399169|complete|auxsubtree f(v0 m2019-01-29 16:23:04.104885 10=10+0) n(v0 rc2019-01-29 16:23:10.722388 b200 10=10+0) hs=10+0,ss=0+0 dirty=10 | child=1 subtree=1 dirty=1 authpin=0 0x55f91c176d00]
2019-01-29 16:23:59.018 7f0db9665700 15 mds.0.cache  subtree bound [dir 0x10000000006 /dir06/ [2,head] auth v=121 cv=0/0 dir_auth=0 state=1611399169|complete|auxsubtree f(v0 m2019-01-29 16:23:04.143036 10=10+0) n(v0 rc2019-01-29 16:23:10.763641 b200 10=10+0) hs=10+0,ss=0+0 dirty=10 | child=1 subtree=1 dirty=1 authpin=0 0x55f91c177200]
2019-01-29 16:23:59.018 7f0db9665700 15 mds.0.cache  subtree bound [dir 0x10000000007 /dir07/ [2,head] auth v=121 cv=0/0 dir_auth=0 state=1611399169|complete|auxsubtree f(v0 m2019-01-29 16:23:04.179958 10=10+0) n(v0 rc2019-01-29 16:23:10.804665 b200 10=10+0) hs=10+0,ss=0+0 dirty=10 | child=1 subtree=1 dirty=1 authpin=0 0x55f91c177700]
2019-01-29 16:23:59.018 7f0db9665700 15 mds.0.cache  subtree bound [dir 0x10000000008 /dir08/ [2,head] auth v=121 cv=0/0 dir_auth=0 state=1611399169|complete|auxsubtree f(v0 m2019-01-29 16:23:04.215531 10=10+0) n(v0 rc2019-01-29 16:23:10.842074 b200 10=10+0) hs=10+0,ss=0+0 dirty=10 | child=1 subtree=1 dirty=1 authpin=0 0x55f91c204000]
2019-01-29 16:23:59.018 7f0db9665700 15 mds.0.cache  subtree bound [dir 0x10000000009 /dir09/ [2,head] auth v=121 cv=0/0 dir_auth=0 state=1611399169|complete|auxsubtree f(v0 m2019-01-29 16:23:04.254493 10=10+0) n(v0 rc2019-01-29 16:23:10.881893 b200 10=10+0) hs=10+0,ss=0+0 dirty=10 | child=1 subtree=1 dirty=1 authpin=0 0x55f91c204500]
```

Do we want to suppress this as well?

rishabh-d-dave commented:
Did rebase and removed the DNM label.

rishabh-d-dave commented:
Did a rebase again.

batrick commented on src/mds/MDCache.cc:
```diff
@@ -12066,6 +12073,12 @@ void MDCache::show_subtrees(int dbl)
     }
   }
 
+  if (depth > SUBTREES_DEPTH_THRESHOLD &&
```
`!force_print &&...`?

batrick commented on src/mds/MDCache.cc:
```diff
@@ -12017,6 +12017,13 @@ void MDCache::show_subtrees(int dbl)
 
+  if (!force_print && subtrees.size() > SUBTREES_COUNT_THRESHOLD &&
+      !g_conf()->subsys.should_gather<ceph_subsys_mds, 25>()) {
+    dout(15) << "number of subtrees = " << subtrees.size() << "; not "
```
`dout(dbl)`

batrick commented on src/mds/MDCache.cc:
```diff
@@ -12066,6 +12073,12 @@ void MDCache::show_subtrees(int dbl)
 
+  if (depth > SUBTREES_DEPTH_THRESHOLD &&
+      !g_conf()->subsys.should_gather<ceph_subsys_mds, 25>()) {
+    dout(15) << "max depth among subtrees = " << depth << "; not printing "
```
`dout(dbl)`

rishabh-d-dave commented:
* added the changes and did a rebase

batrick commented on src/mds/MDCache.cc:
```diff
@@ -12038,6 +12038,13 @@ void MDCache::show_subtrees(int dbl)
 
+  if (!force_print && subtrees.size() > SUBTREES_COUNT_THRESHOLD &&
+      !g_conf()->subsys.should_gather<ceph_subsys_mds, 25>()) {
+    dout(dbl) << "number of subtrees = " << subtrees.size() << "; not "
```
oops, `dout(dout(ceph::dout::need_dynamic(dbl)))`

rishabh-d-dave replied:
Made the change at [other instance](https://github.com/ceph/ceph/pull/26056/files#diff-2259d0a61fc0c9b2d2be3e40ade57303R12044) too.

batrick commented:
ping @rishabh-d-dave

rishabh-d-dave commented:
Sorry about the delay.  PR #26477 turned out to be bigger than I had expected when I begun with it. Besides, I was a little stuck due to changes I made last time (replaciing 15 in dout() by dbl). The changes didn't compile for me when I continued with PR yesterday. This confused me for two reasons; first, I vaguely remember build completing successfully and second, is not writing 15 more or less same as writing dbl since both are int vaues? I haven't understood what the build error message ([attached](https://github.com/ceph/ceph/files/2911802/error-with-dbl-instead-of-15.log)) tries to say. However, after the making the latest change, the build completed sucessfully, so it's not a blocking issue anymore.

batrick commented on src/mds/MDBalancer.cc:
```diff
@@ -150,11 +150,20 @@ void MDBalancer::handle_export_pins(void)
+    dout(15) << "number of auth trees = " << authsubs.size() << "; not "
+		"printing auth trees" << dendl;
+    return;
+  }
```
````
2019-03-01 19:39:07.919 1aa69700 20 mds.6.bal export_pin_queue size=0
2019-03-01 19:39:07.919 1aa69700 15 mds.6.bal number of auth trees = 34; not printing auth trees
2019-03-01 19:39:07.919 1aa69700 15 mds.6.bal tick last_sample now 3644.81s
2019-03-01 19:39:07.919 1aa69700 15 mds.6.cache show_subtrees
2019-03-01 19:39:07.920 1aa69700 10 mds.6.cache |__________ 6    auth [dir 0x106 ~mds6/ [2,head] auth v=11 cv=1/1 dir_auth=6 state=1610612737|complete f(v0 10=0+10) n(v1 rc2019-03-01 19:37:27.978749 b4194304 11=1+10)/n(v1 rc2019-03-01 19:36:30.306667 b1973 11=1+10) hs=10+0,ss=0+0 dirty=1 | child=1 subtree=1 dirty=1 authpin=0 0x1b381170]
````

From: /ceph/teuthology-archive/pdonnell-2019-03-01_18:13:24-multimds-wip-pdonnell-testing-20190301.150213-distro-basic-smithi/3654417/remote/smithi178/log/ceph-mds.e.log.gz

This isn't right. You are skipping the debug prints in the following lines but you're also skipping the call to `export_dir`.

rishabh-d-dave replied:
Fixed.

rishabh-d-dave commented:
change:

* incorporated changes requested

rishabh-d-dave commented:
@batrick 

FYKI, the issue I wanted to talk about today on IRC got resolved. :)

Also, can you please reply on this question - https://github.com/ceph/ceph/pull/26056#discussion_r250905792 ?

rishabh-d-dave commented:
change:

* nit: moved `mds->mdcache->show_subtrees(10, true);` in `MDBalancer::tick()` from one commit to another since it not about auth subtress

rishabh-d-dave commented:
@batrick 
I found this part of log[(attached)](https://github.com/ceph/ceph/files/2938003/pin-set-nonexisting-mds.log) a bit repetitive like the ones we are trying to omit here. This happened when at some point today while working on this PR, I did `sudo setfattr -n ceph.dir.pin -v 2 /mnt/cephfs1/dir$i` instead of `sudo setfattr -n ceph.dir.pin -v 2 /mnt/cephfs1/dir$i` on a cluster with 2 MDSs.

On a quick overview, I think at least some of the message can be omitted or reduced into something more simpler or succinct, but I can't be sure since this isn't as straightforward as printing auth subtrees. Let me know what you think.

batrick commented on src/mds/MDBalancer.cc:
```diff
@@ -150,11 +150,26 @@ void MDBalancer::handle_export_pins(void)
+
     if (export_pin >= 0 && export_pin != mds->get_nodeid()) {
-      dout(10) << "exporting auth subtree " << *cd->inode << " to " << export_pin << dendl;
+      dout(25) << "exporting auth subtree " << *cd->inode << " to " <<
```
I think we can just drop this debug log message. It's clear what happens due to the early messages in `Migrator::export_dir`.

batrick commented:
> On a quick overview, I think at least some of the message can be omitted or reduced into something more simpler or succinct, but I can't be sure since this isn't as straightforward as printing auth subtrees. Let me know what you think.

I think we can bump the debugging level for "dir is export pinned" and move the check to the beginning of the method so we avoid other messages.

rishabh-d-dave commented:
changes:

* incorporated changes requested
* bumped up the message for "dir is export pinned" debug message and moved the check at the beginning of the method

rishabh-d-dave commented:
changes:

* rebased and resolved the merge confilict
